br/pkg/lightning/backend/local/region_job.go

```diff
@@ -619,10 +661,141 @@ func (j *regionJob) fixIngestError(
 		j.convertStageTo(needRescan)
 		return false, nil
 	case errPb.DiskFull != nil:
+		j.lastRetryableErr = common.ErrKVIngestFailed.GenWithStack(errPb.GetMessage())
+
 		return false, errors.Errorf("non-retryable error: %s", resp.GetError().GetMessage())
 	}
 	// all others doIngest error, such as stale command, etc. we'll retry it again from writeAndIngestByRange
 	j.lastRetryableErr = common.ErrKVIngestFailed.GenWithStack(resp.GetError().GetMessage())
 	j.convertStageTo(regionScanned)
 	return false, nil
 }
+
+type regionJobRetryHeap []*regionJob
+
+func (h regionJobRetryHeap) Len() int {
+	return len(h)
+}
+
+func (h regionJobRetryHeap) Less(i, j int) bool {
+	return h[i].waitUntil.Before(h[j].waitUntil)
+}
+
+func (h regionJobRetryHeap) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *regionJobRetryHeap) Push(x any) {
+	*h = append(*h, x.(*regionJob))
+}
+
+func (h *regionJobRetryHeap) Pop() any {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+// regionJobRetryer is a concurrent-safe queue holding jobs that need to put
+// back later, and put back when the regionJob.waitUntil is reached. It maintains
+// a heap of jobs internally based on the regionJob.waitUntil field.
+type regionJobRetryer struct {
+	q         regionJobRetryHeap
+	qMu       sync.Mutex
+	putBackCh chan<- *regionJob
+	reload    chan struct{}
+	done      chan struct{}
+	closed    bool
+	closeMu   sync.Mutex
+}
+
+func newRegionJobRetryer(putBackCh chan<- *regionJob) *regionJobRetryer {
+	return &regionJobRetryer{
+		q:         make([]*regionJob, 0, 16),
+		putBackCh: putBackCh,
+		reload:    make(chan struct{}, 1),
+		done:      make(chan struct{}),
+	}
+}
+
+func (q *regionJobRetryer) run(ctx context.Context) {
+	var (
+		front     *regionJob
+		toPutBack *regionJob
+	)
+
+	for {
+		q.qMu.Lock()
+		if len(q.q) > 0 {
+			front = q.q[0]
+		} else {
+			front = nil
+		}
+		q.qMu.Unlock()
+
+		switch {
+		case toPutBack != nil:
+			select {
+			case <-ctx.Done():
+				return
+			case <-q.done:
```

Comment:
if we done `close`, but `toPutBack` is taken out, so we might miss one jobWg.Done

Reply:
https://github.com/pingcap/tidb/pull/42780/commits/20fc8f5d064dcd70714373f08ca83bea43ada28f